### PR TITLE
[Design] Use DictionaryKeyResolver for ExpandoObjectAdapter

### PR DESCRIPTION
**Pull request by jbagga**
Addresses https://github.com/aspnet/JsonPatch/issues/102
@rynowak Let's talk about what else needs to change. We may also need tests that mock the IContractResolver to return a contract with a custom naming strategy.

### CODE REVIEW

File: test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/AddOperationTests.cs [hunk 1]
```diff
@@ -396,7 +396,7 @@ public void ShouldReplacePropertyWithDifferentCase()
 
             deserialized.ApplyTo(doc);
 
-            Assert.Equal("B", doc.StringProperty);
+            Assert.Equal("A", doc.StringProperty);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/AddOperationTests.cs]: This test should be updated to verify that both keys are present.

File: test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/RemoveOperationTests.cs [hunk 2]
```diff
@@ -80,7 +80,7 @@ public void RemovePropertyFromExpandoObject()
         }
 
         [Fact]
-        public void RemovePropertyFromExpandoObjectMixedCase()
+        public void RemovePropertyFromExpandoObjectMixedCaseThrowsPathNotFoundException()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/RemoveOperationTests.cs]: You can break up these names with underscores to make it more readable. 

`RemoveProperty_FromExpandoObject_MixedCase_ThrowsPathNotFoundException`

File: test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/RemoveOperationTests.cs [hunk 3]
```diff
@@ -125,7 +126,7 @@ public void RemoveNestedPropertyFromExpandoObject()
         }
 
         [Fact]
-        public void RemoveNestedPropertyFromExpandoObjectMixedCase()
+        public void RemoveNestedPropertyFromExpandoObjectMixedCaseThrowsPathNotFoundException()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/RemoveOperationTests.cs]: It would be good to have a test here with a non-default naming strategy.

**Comment by jbagga** [comment]
@rynowak Updated with tests for a custom `NamingStrategy`. Also some effort to secure Pranav points

File: src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs [hunk 4]
```diff
@@ -406,10 +391,8 @@ public void Copy(Operation operation, object objectToApplyTo)
             var parsedPath = new ParsedPath(fromLocation);
             var visitor = new ObjectVisitor(parsedPath, ContractResolver);
 
-            IAdapter adapter;
             var target = objectToGetValueFrom;
-            string errorMessage;
-            if (!visitor.TryVisit(ref target, out adapter, out errorMessage))
+            if (!visitor.TryVisit(ref target, out var adapter, out var errorMessage))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs]: /cc @pranavkm

File: test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/AddOperationTests.cs (hunk 1 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.JsonPatch.Test/Dynamic/AddOperationTests.cs]: Looks good now 👍

File: src/Microsoft.AspNetCore.JsonPatch/Internal/ExpandoObjectAdapter.cs [hunk 5]
```diff
@@ -16,10 +16,10 @@ public class ExpandoObjectAdapter : IAdapter
             object value,
             out string errorMessage)
         {
+            var contract = (JsonDictionaryContract)contractResolver.ResolveContract(target.GetType());
+            var key = contract.DictionaryKeyResolver(segment);
             var dictionary = (IDictionary<string, object>)target;
 
-            var key = dictionary.GetKeyUsingCaseInsensitiveSearch(segment);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/Internal/ExpandoObjectAdapter.cs]: Can this method be deleted now?

**Comment by jbagga** [comment]
@rynowak Removed the method. I think @Eilon is aware of it. I can confirm before I merge this.

File: src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs]: 💯 <sup> 💯 </sup>

**Pull request merged**